probe-rs/src/core/core_state.rs

```diff
@@ -1,37 +1,200 @@
+
+#[derive(Debug)]
+pub(crate) struct CombinedCoreState {
+    pub(crate) core_state: CoreState,
+
+    pub(crate) specific_state: SpecificCoreState,
+
+    pub(crate) id: usize,
+}
+
+impl CombinedCoreState {
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    pub fn core_type(&self) -> CoreType {
+        self.specific_state.core_type()
+    }
+
+    pub(crate) fn attach_arm<'probe>(
+        &'probe mut self,
+        arm_interface: &'probe mut Box<dyn ArmProbeInterface>,
+    ) -> Result<Core<'probe>, Error> {
+        let memory = arm_interface.memory_interface(self.arm_memory_ap())?;
+
+        let (options, debug_sequence) = match &self.core_state.core_access_options {
+            ResolvedCoreOptions::Arm { options, sequence } => (options, sequence.clone()),
+            ResolvedCoreOptions::Riscv { .. } => {
+                return Err(Error::UnableToOpenProbe(
+                    "Core architecture and Probe mismatch.",
+                ))
+            }
+        };
+
+        Ok(match &mut self.specific_state {
+            SpecificCoreState::Armv6m(s) => Core::new(
+                crate::architecture::arm::armv6m::Armv6m::new(memory, s, debug_sequence, self.id)?,
+            ),
+            SpecificCoreState::Armv7a(s) => {
+                Core::new(crate::architecture::arm::armv7a::Armv7a::new(
+                    memory,
+                    s,
+                    options.debug_base.expect("base_address not specified"),
+                    debug_sequence,
+                    self.id,
+                )?)
+            }
+            SpecificCoreState::Armv7m(s) | SpecificCoreState::Armv7em(s) => Core::new(
+                crate::architecture::arm::armv7m::Armv7m::new(memory, s, debug_sequence, self.id)?,
+            ),
+            SpecificCoreState::Armv8a(s) => {
+                Core::new(crate::architecture::arm::armv8a::Armv8a::new(
+                    memory,
+                    s,
+                    options.debug_base.expect("base_address not specified"),
+                    options.cti_base.expect("cti_address not specified"),
+                    debug_sequence,
+                    self.id,
+                )?)
+            }
+            SpecificCoreState::Armv8m(s) => Core::new(
+                crate::architecture::arm::armv8m::Armv8m::new(memory, s, debug_sequence, self.id)?,
+            ),
+            _ => {
+                return Err(Error::UnableToOpenProbe(
+                    "Core architecture and Probe mismatch.",
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn enable_arm_debug(
+        &self,
+        interface: &mut dyn ArmProbeInterface,
+    ) -> Result<(), Error> {
+        let (sequence_handle, arm_core_access_options) = match &self.core_state.core_access_options
+        {
+            ResolvedCoreOptions::Arm { sequence, options } => (sequence, options),
+            ResolvedCoreOptions::Riscv { .. } => {
+                todo!()
```

Comment:
I think this should rather be a panic explaining that we hit a bug.

Reply:
I would really prefer if we didn't panic. It causes very ungraceful behaviour under the VSCode extension and debugger. Basically, the session just disappears and the user has no idea of what happens. Can we throw an error from the lib level, so  that the tool can decide to message the user or panic?

Reply:
I think we should really look into a panic catchall (https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) for vscode.

Generally this just makes the library design much much worse if we throw dynamic errors everywhere. This panic should never happen if we do not mess up the logic. Introducing error variants for every case is really cluttery unfortunately and is actually discouraged I think.

I always did this in the past and came to embrace the usefulness of panic :)